Enhanced with config for time between retries and isJobRunnable function

robnewton opened the pull request:
Added support for the following capabilities:

- Configurable minimum time (in milliseconds) between attempts. Set this as `minimumMillisBetweenAttempts` in the worker options object.
- Configurable `isJobRunnable` function which is called after initially fetching a job from Realm to check if it should be processed. This has access to the entire move, it's payload, the network connection status, as well as a realm object for referencing other db items. Also an example implementation is included in the readme for requiring a minimum wait time after job creation before working it.

An implementation has started for `failureBehavior` but at this point it may not be necessary. Left it in for now so we can test and easily flesh it out if we need to. Default is `standard` but if you set it to anything other than that in the worker options then it will break failure handling right now.

robnewton committed:
Added queue.start() session tracking on jobs

robnewton committed:
Single failed job per queue.start() session

robnewton committed:
Added session filtering to main query

robnewton committed:
Increment minor version

robnewton committed:
Corrected model issues and query filter

robnewton committed:
Fixed query filter

robnewton committed:
Added configurable retries and min duration between

robnewton committed:
filter issues wip

robnewton committed:
Fixed bug with excluded jobs

robnewton committed:
Fixed with worker specific filters

robnewton committed:
isRunnable + onSkipped + minimumMillisBetweenAttempts

robnewton committed:
isJobRunnable takes full job now

robnewton committed:
Updated readme

robnewton committed:
Removed notes

robnewton committed:
Removed the unused session field

robnewton committed:
Fixed default isJobRunnable sig

robnewton committed:
Fixed default runnable function

bescott1 commented on Models/Queue.js:
```diff
@@ -217,15 +218,15 @@ export class Queue {
    * @param queueLifespanRemaining {number} - The remaining lifespan of the current queue process (defaults to indefinite).
    * @return {promise} - Promise resolves to an array of job(s) to be processed next by the queue.
    */
   async getConcurrentJobs(queueLifespanRemaining = 0) {
```
this function has grown quite a bit. Consider breaking it up into multiple functions that break up the different logistical steps

bescott1 commented on Models/Queue.js:
```diff
@@ -234,9 +235,46 @@ export class Queue {
       const initialQuery = (queueLifespanRemaining)
-        ? '(active == FALSE AND failed == null AND timeout > 0 AND timeout < ' + timeoutUpperBound + ') OR (active == FALSE AND failed == null AND timeout > 0 AND timeout < ' + timeoutUpperBound + ')'
-        : '(active == FALSE AND failed == null) OR (active == TRUE && failed == null)';
+        ? `
+           (active == FALSE                AND
+            failed == null                 AND
+            (${workerFilters?.join(' OR ')}) AND
+            timeout > 0                    AND
+            timeout < ${timeoutUpperBound})
+          OR (active == FALSE              AND
+            failed == null                 AND
+            (${workerFilters?.join(' OR ')}) AND
+            timeout > 0                    AND
+            timeout < ${timeoutUpperBound})`
+
+        : `
+           (active == FALSE                AND
+            (${workerFilters?.join(' OR ')}) AND
+            failed == null)
+          OR (active == TRUE               AND
+            (${workerFilters?.join(' OR ')}) AND
+            failed == null)`;
```
break queries outside of the function

bescott1 commented on Models/Queue.js:
```diff
@@ -251,37 +289,70 @@ export class Queue {
         const allRelatedJobsQuery = (queueLifespanRemaining)
-          ? '(name == "'+ nextJob.name +'" AND active == FALSE AND failed == null AND timeout > 0 AND timeout < ' + timeoutUpperBound + ') OR (name == "'+ nextJob.name +'" AND active == FALSE AND failed == null AND timeout > 0 AND timeout < ' + timeoutUpperBound + ')'
-          : '(name == "'+ nextJob.name +'" AND active == FALSE AND failed == null) OR (name == "'+ nextJob.name +'" AND active == TRUE AND failed == null)';
+          ? `(name == "${nextJob.name}"   AND
```
Same break queries into another file

bescott1 commented on Models/Queue.js:
```diff
@@ -326,39 +397,55 @@ export class Queue {
+
+      switch (failureBehavior) {
+        case 'standard':
+          this.realm.write(() => {
```
break the inside of the case statement into a function so it's easier to validate the syntax of the swtich statement and easier to read

bescott1 reviewed (commented):
Just did a quick passthrough and added comments. May have more later

robnewton committed:
Added get job function

robnewton committed:
Fixed getJob

robnewton committed:
Status change observer

robnewton committed:
Corrected comments

robnewton committed:
Added some comments

robnewton committed:
onQueueJobChanged event handler

robnewton committed:
Added skip tracking

robnewton committed:
Updated package name prefix

robnewton committed:
Try that again

robnewton committed:
Added publish registry of github @hopdrive

robnewton committed:
Added get workers as array

dartushd committed:
Update version 2.4.0

dartushd committed:
Merge branch 'feature/enhanced-error-handling' of github.com:hopdrive/react-native-queue into feature/delete-single-job

dartushd committed:
weird name change

robnewton committed:
Merge pull request #3 from hopdrive/feature/delete-single-job

Delete jobs by id and all failed jobs

ian-wd committed:
build: bump realm to latest v12 (#26)

robnewton committed:
Upgraded package to access realm path

jyann committed:
TCS-1288 switch to npm (#27)

* TCS-1288 build: switch from yarn to npm

* TCS-1288 build: drop networ-concurrency option

dependabot[bot] committed:
build(deps): bump ajv and eslint (#28)

* build(deps): bump ajv and eslint

Bumps [ajv](https://github.com/ajv-validator/ajv) to 6.12.6 and updates ancestor dependency [eslint](https://github.com/eslint/eslint). These dependencies need to be updated together.


Updates `ajv` from 5.5.2 to 6.12.6
- [Release notes](https://github.com/ajv-validator/ajv/releases)
- [Commits](https://github.com/ajv-validator/ajv/compare/v5.5.2...v6.12.6)

Updates `eslint` from 4.19.1 to 8.56.0
- [Release notes](https://github.com/eslint/eslint/releases)
- [Changelog](https://github.com/eslint/eslint/blob/main/CHANGELOG.md)
- [Commits](https://github.com/eslint/eslint/compare/v4.19.1...v8.56.0)

---
updated-dependencies:
- dependency-name: ajv
  dependency-type: indirect
- dependency-name: eslint
  dependency-type: direct:development
...

Signed-off-by: dependabot[bot] <support@github.com>

* chore: lint eslintrc

* build: add "lint:fix" command

---------

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Connor Tumbleson <connor@sourcetoad.com>

robnewton committed:
Merge branch 'master' into feature/enhanced-error-handling

dartushd committed:
Merge branch 'master' into feature/enhanced-error-handling
